content/blog/2019-12-18-polynomial/index.md

```diff
@@ -0,0 +1,197 @@
+algorithm. By trying out different combination of rules, it can find the best
+implementation on the target platform. 
+
+## Efficient Polynomial Multiplication Algorithm
+
+Inspired from Spiral, I thought that we could similarly write down the rules to 
+break down polynomial multiplication, and we can use these rules to break down polynomial
+multiplication into smaller ones. As we can see in this section, there are
+different rules we can use to recursively break down the polynomial.
+
+
+### Schoolbook Multiplication
+Suppose we want to multiply two polynomials $f(x)$ and $g(x)$, where
+  $$
+  f(x) = a_1(x) t + a_0(x)
+  $$
+  $$
+  g(x) = b_1(x) t + b_0(x)
+  $$
+
+  Then schoolbook method is to compute 
+  $$
+  f(x)g(x) = c_2(x) t^2 + c_1 t + c_0(x)
+  $$
+  where
+  $$ c_2(x) = a_1(x)b_1(x) $$
+  $$ c_1(x) = a_1(x)b_0(x) + a_0(x)b_1(x)$$
+  $$ c_0(x) = a_0(x)b_0(x) $$.
+  It can break down a large polynomial multiplications into 4 smaller polynomial
+  multiplications.
+
+### Karatsuba
+We can break the polynomial $f(x)$ and $g(x)$ to higher degree part
+and lower degree part as follows:
+
+  $$
+  f(x) = a_1(x) t + a_0(x)
+  $$
+  $$
+  g(x) = b_1(x) t + b_0(x)
+  $$
+  , where $t = x^{\frac{n}{2}}$. 
+  Then the multiplication can be calculated as follows:
+
+  $$
+  f(x)g(x) = c_2(x) t^2 + c_1 t + c_0(x)
+  $$
+
+  where 
+  $$ c_2(x) = a_1(x)b_1(x) $$
+  $$ c_0(x) = a_0(x)b_0(x) $$
+
+  $$
+  c_1(x) = (a_1(x) + a_0(x)) (b_1(x) +b_0(x)) -c_0(x)-c_2(x)
+  $$
+
+  It uses 3 multiplications while the schoolbook method would use 4.
+  The idea can be further generalized to a k-way split, and it is called Toom-k.
+
+### Toom-k
+
+Lets take Toom-3 as an example. First we divide the polynomial $f(x)$ and $g(x)$ into 3 parts.
+
+$$
+f(x) = a_2(x) t^2 + a_1(x) t + a_0(x)
+$$
+
+$$
+g(x) = b_2(x) t^2 + b_1(x) t + b_0(x)
+$$
+
+where $t = x^{\frac{n}{3}}$. 
+Similarly to Karatsuba method, in this case, the number of
+multiplication can be reduced to 5, instead of 9 in the case of schoolbook
+method and asymptotically better than both Karatsuba method and schoolbook
+method. This method can be generalized to $k$-way split.
+
+### Unable to decouple the algorithm from the order of computation
+However, I am unable to find a way to decouple the above algorithm from the
+order of computation while still be easy to configure and be useful. As a
+result, it makes this project unsuccessful.
+
+## Implementation
+
+The goal of this project is to build a system whose input is n, the degree of
+the input polynomial, and output a optimized AVX2 code. To make this happen, we
+need to at least take into account different algorithms and different order of
+computation. Without the ability to specify the order of computation easily, I
+do not accomplish the goal I set out to do. Nevertheless, I still implemented
+different algorithm as the break down rules and generate C++ code with AVX2 for
```

Comment:
What is "C++ code with AVX2"? Did you use GCC's vector intrinsics, for example? Or inline assembly?